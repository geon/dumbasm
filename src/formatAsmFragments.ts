import type { AsmFragment } from "./parsers/parseAsmLine";
import type { Operand } from "./parsers/addressingModeCategoryParsers";
import type {
	GetAddressingModeCategory,
	ParsedAddressingModeCategory,
	ParsedMos6502Instruction,
} from "./parsers/parseMos6502Instruction";
import type { Mos6502AddressingMode } from "./mos6502Instructions";

export function formatAsmFragments(fragments: readonly AsmFragment[]): string {
	return fragments.map(formatAsmFragment).join("\n");
}

function formatAsmFragment(fragment: AsmFragment): string {
	switch (fragment.type) {
		case "label": {
			return fragment.value + ":";
		}

		case "directive": {
			return fragment.value;
		}

		case "instruction": {
			return formatInstruction(fragment.value);
		}

		// Exclude this case from code coverage, since it is only for type safety.
		/* v8 ignore next 3 */
		default: {
			throw new Error(fragment satisfies never);
		}
	}
}

function formatInstruction(instruction: ParsedMos6502Instruction): string {
	return instruction.addressingMode.type === "implied"
		? instruction.mnemonic
		: instruction.mnemonic +
				" " +
				formatAddressingMode(instruction.addressingMode);
}

function formatAddressingMode(
	operand: ParsedAddressingModeCategory<
		Exclude<GetAddressingModeCategory<Mos6502AddressingMode>, "implied">
	>,
): string {
	switch (operand.type) {
		case "immediate": {
			return `#${formatNumber(operand.value)}`;
		}

		case "preIndexedIndirect": {
			return `(${formatOperand(operand.value.operand)},X)`;
		}

		case "postIndexedIndirect": {
			return `(${formatOperand(operand.value.operand)}),Y`;
		}

		case "indirect": {
			return `(${formatOperand(operand.value)})`;
		}

		case "indexed": {
			return `${formatOperand(operand.value.operand)},${operand.value.index}`;
		}

		case "accumulator": {
			return `A`;
		}

		case "address": {
			return `${formatOperand(operand.value)}`;
		}

		// Exclude this case from code coverage, since it is only for type safety.
		/* v8 ignore next 3 */
		default: {
			throw new Error(operand satisfies never);
		}
	}
}

function formatOperand(operand: Operand): string {
	switch (operand.type) {
		case "number": {
			return formatNumber(operand.value);
		}

		case "identifier": {
			return operand.value;
		}

		// Exclude this case from code coverage, since it is only for type safety.
		/* v8 ignore next 3 */
		default: {
			throw new Error(operand satisfies never);
		}
	}
}

function formatNumber(number: number): string {
	return `$${number.toString(16)}`;
}
