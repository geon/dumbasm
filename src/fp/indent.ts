const indentChar = "\t";

export function indent(_text: string): string {
	return indentChar;
}
