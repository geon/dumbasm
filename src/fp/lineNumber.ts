export function getLineBeginIndex(input: string, fromIndex: number) {
	if (fromIndex === 0) {
		return 0;
	}

	const lastIndex = input.lastIndexOf("\n", fromIndex - 1);

	return lastIndex + 1;
}

export function getLineEndIndex(input: string, fromIndex: number) {
	const index = input.indexOf("\n", fromIndex);
	if (index === -1) {
		return undefined;
	}

	return index;
}
