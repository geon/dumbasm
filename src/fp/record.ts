export function mapRecord<TKey extends string, TIn, TOut>(
	record: Readonly<Record<TKey, TIn>>,
	transform: (value: TIn, key: TKey) => TOut,
): Record<TKey, TOut> {
	return Object.fromEntries(
		Object.entries(record).map(([key, value]) => [
			key,
			transform(value as TIn, key as TKey),
		]),
	) as Readonly<Record<TKey, TOut>>;
}

// https://stackoverflow.com/questions/69019873/how-can-i-get-typed-object-entries-and-object-fromentries-in-typescript
export function objectFromEntries<
	const T extends ReadonlyArray<readonly [PropertyKey, unknown]>,
>(entries: T): { [K in T[number] as K[0]]: K[1] } {
	return Object.fromEntries(entries) as { [K in T[number] as K[0]]: K[1] };
}
