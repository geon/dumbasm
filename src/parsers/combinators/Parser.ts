export type ParseResult<T> =
	| {
			readonly type: "result";
			readonly consumed: number;
			readonly parsed: T;
	  }
	| {
			readonly type: "error";
			readonly fromIndex: number;
			readonly message: string;
	  };

export type ParseError = Extract<ParseResult<unknown>, { type: "error" }>;

export function createParseResult<T>(
	consumed: number,
	parsed: T,
): Extract<ParseResult<T>, { type: "result" }> {
	return { type: "result", consumed, parsed };
}

export function createParseError(
	fromIndex: number,
	message: string,
): ParseError {
	return { type: "error", fromIndex, message };
}

export type ParserArgs = readonly [input: string, fromIndex: number];

export type Parser<T> = (...args: ParserArgs) => ParseResult<T>;

export function parsingFailed(
	parseResult: ParseResult<unknown>,
): parseResult is ParseError {
	return parseResult.type === "error";
}
