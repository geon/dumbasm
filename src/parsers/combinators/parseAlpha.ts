import { parseAlternatives } from "./parseAlternatives.js";
import { parseChar } from "./parseChar.js";
import type { Parser } from "./Parser.js";
import { parseWithErrorMessage } from "./parseWithErrorMessage.js";

const alphas = [
	[...Array(26)].map((_, x) => [
		String.fromCharCode("a".charCodeAt(0) + x),
		String.fromCharCode("A".charCodeAt(0) + x),
	]),
].flat();

export const parseAlpha: Parser<string> = parseWithErrorMessage(
	"Expected a-z or A-Z.",
	parseAlternatives(alphas.flat().map(parseChar)),
);
