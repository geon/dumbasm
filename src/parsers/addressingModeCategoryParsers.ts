import { parseChar } from "./combinators/parseChar.js";
import { parseKeyed } from "./combinators/parseKeyed.js";
import { parseMonad } from "./combinators/parseMonad.js";
import { parseNothing } from "./combinators/parseNothing.js";
import { type Parser } from "./combinators/Parser.js";
import { parseSequence } from "./combinators/parseSequence.js";
import { parseSequenceIndex } from "./combinators/parseSequenceIndex.js";
import { parseString } from "./combinators/parseString.js";
import { parseIdentifier } from "./parseIdentifier.js";
import { parseNumber } from "./parseNumber.js";

export type Operand =
	| {
			type: "number";
			value: number;
	  }
	| {
			type: "identifier";
			value: string;
	  };

const parseOperand: Parser<Operand> = parseKeyed({
	number: parseNumber,
	identifier: parseIdentifier,
});

export const addressingModeCategoryParsers = {
	implied: parseNothing(undefined),
	accumulator: parseMonad(parseIdentifier, (identifier, { error, result }) =>
		identifier.toUpperCase() === "A"
			? result(undefined)
			: error("Expected A for accumulator."),
	),
	immediate: parseSequenceIndex(1, [parseChar("#"), parseNumber]),
	address: parseOperand,
	indexed: <T>(parseIndex: Parser<T>) =>
		parseMonad(
			parseSequence([parseOperand, parseChar(","), parseIndex]),
			([operand, , index], { result }) => result({ operand, index }),
		),
	indirect: parseSequenceIndex(1, [
		parseChar("("),
		parseOperand,
		parseChar(")"),
	]),
	preIndexedIndirect: <T>(parseIndex: Parser<T>) =>
		parseMonad(
			parseSequence([
				parseChar("("),
				parseOperand,
				parseChar(","),
				parseIndex,
				parseChar(")"),
			]),
			([, operand, , index], { result }) => result({ operand, index }),
		),
	postIndexedIndirect: <T>(parseIndex: Parser<T>) =>
		parseMonad(
			parseSequence([
				parseChar("("),
				parseOperand,
				parseString("),"),
				parseIndex,
			]),
			([, operand, , index], { result }) => result({ operand, index }),
		),
};
