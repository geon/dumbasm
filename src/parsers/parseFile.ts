import { parseEof } from "./combinators/parseEof.js";
import { parseMonad } from "./combinators/parseMonad.js";
import { type ParserArgs, type ParseResult } from "./combinators/Parser.js";
import { parseWithErrorMessage } from "./combinators/parseWithErrorMessage.js";
import { type AsmFragment } from "./parseAsmLine.js";

export type ParsedFile = readonly AsmFragment[];

export function parseFile(...args: ParserArgs): ParseResult<ParsedFile> {
	return parseWithErrorMessage(
		"SYNTAX ERROR",
		parseMonad(parseEof, (_, { result }) => result([])),
	)(...args);
}
