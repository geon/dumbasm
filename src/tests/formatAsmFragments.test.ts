import { suite, test, expect, assert } from "vitest";
import { formatAsmFragments } from "../formatAsmFragments";
import { parseFile } from "../parsers/parseFile";
import { parsingFailed, type ParseError } from "../parsers/combinators/Parser";
import { asmSamples } from "../parsers/tests/asm-samples";
import { getLineCol } from "../fp/lineNumber";

testExamples("formatAsmFragments", [
	{
		name: "hello world",
		input: asmSamples.helloWorld,
	},
	{
		name: "missing modes",
		input: asmSamples.addressingModesMissingInHelloWorld,
	},
]);

export function testExamples(
	suiteName: string,
	examples: readonly {
		readonly name: string;
		readonly input: string;
	}[],
) {
	suite(suiteName, () => {
		for (const example of examples) {
			test(example.name, () => {
				const parsed = parseFile(example.input, 0);

				if (parsingFailed(parsed)) {
					assert.fail(formatError(parsed, example.input));
				} else {
					const asmFragments = parsed.parsed.map((fragment) => {
						if (
							fragment.type === "variableDeclaration" ||
							fragment.type === "scope"
						) {
							throw new Error("Not implemented.");
						}
						return fragment;
					});

					expect(formatAsmFragments(asmFragments)).toMatchSnapshot();
				}
			});
		}
	});
}

function formatError(parsed: ParseError, input: string): string {
	const { lineBeginIndex, lineEndIndex, columnNumber, lineNumber } = getLineCol(
		input,
		parsed.fromIndex,
	);

	const line = input.slice(lineBeginIndex, lineEndIndex);
	const columnIndex = columnNumber - 1;
	const arrow =
		Array.from(
			{
				length: columnIndex,
			},
			(_, index) => (line[index] === "\t" ? "\t" : " "),
		).join("") + "^";

	return `Parse Error on line ${lineNumber}:${columnNumber}\n${parsed.message}\n${line}\n${arrow}`;
}
