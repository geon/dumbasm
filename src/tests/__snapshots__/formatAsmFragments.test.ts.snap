// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatAsmFragments > hello world 1`] = `
"ldy #$0
loop:
lda message,Y
beq done
jsr $ffd2
iny
jmp loop
done:
rts
message:
.byte 147 ; clear
.text "HELLO WORLD"
.byte 0 ; null terminator"
`;

exports[`formatAsmFragments > missing modes 1`] = `
"rol A
jmp ($ab12)
lda ($ab12),Y
lda ($ab12,X)"
`;
